apps/rush-lib/src/logic/pnpm/PnpmfileShim.ts

```diff
@@ -1,53 +1,72 @@
-    readPackage?: (pkg: IPackageJson, context: any) => IPackageJson;
-  };
+// Initialize all external aspects of the pnpmfile shim. When using the shim, settings
+// are always expected to be available. The rest can be considered additional and are
+// not guaranteed at runtime. Init must be called before running any hook that depends
+// on a resource obtained from or related to the settings, and will require modules
+// once so they aren't repeatedly required in the hook functions.
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+function init(context: IPnpmfileContext | any): IPnpmfileContext {
+  // Sometimes PNPM may provide us a context arg that doesn't fit spec, ex.:
+  // https://github.com/pnpm/pnpm/blob/97c64bae4d14a8c8f05803f1d94075ee29c2df2f/packages/get-context/src/index.ts#L134
+  // So we need to ensure the context format before we move on
+  if (typeof context !== 'object' || Array.isArray(context)) {
+    context = {
+      log: (message: string) => {},
+      originalContext: context
+    } as IPnpmfileContext;
+  }
+  if (!settings) {
+    // Initialize the settings from file
+    if (!context.pnpmfileShimSettings) {
+      context.pnpmfileShimSettings = require('./pnpmfileSettings.json');
+    }
+    settings = context.pnpmfileShimSettings!;
+  } else if (!context.pnpmfileShimSettings) {
+    // Reuse the already initialized settings
+    context.pnpmfileShimSettings = settings;
+  }
+  if (!clientPnpmfile && settings.clientPnpmfilePath) {
+    clientPnpmfile = require(settings.clientPnpmfilePath);
+  }
+  if (!semver && settings.semverPath) {
+    semver = require(settings.semverPath);
+  }
+  return context as IPnpmfileContext;
 }
 
-// Load in the generated settings file
-const pnpmfileSettings: IPnpmfileShimSettings = require('./pnpmfileSettings.json');
-// We will require semver from this path on disk, since this is the version of semver shipping with Rush
-const semver: typeof TSemver = require(pnpmfileSettings.semverPath);
-// Only require the client pnpmfile if requested
-const clientPnpmfile: IPnpmfile | undefined = pnpmfileSettings.useClientPnpmfile
-  ? require('./clientPnpmfile')
-  : undefined;
-
 // Set the preferred versions on the dependency map. If the version on the map is an allowedAlternativeVersion
 // then skip it. Otherwise, check to ensure that the common version is a subset of the specified version. If
 // it is, then replace the specified version with the preferredVersion
-function setPreferredVersions(dependencies?: { [dependencyName: string]: string }): void {
+function setPreferredVersions(dependencies: { [dependencyName: string]: string } | undefined): void {
   for (const name of Object.keys(dependencies || {})) {
-    if (pnpmfileSettings.allPreferredVersions.hasOwnProperty(name)) {
-      const preferredVersion: string = pnpmfileSettings.allPreferredVersions[name];
+    if (settings.allPreferredVersions?.hasOwnProperty(name)) {
```

Comment:
Can we convert this to a `Map<string, string>`? `new Map(Object.entries(obj))` will do the job.